File: cinema/src/main/java/com/cinema/user/service/UserService.java
package com.cinema.user.service;

import com.cinema.shared.exception.BusinessException;
import com.cinema.shared.exception.ErrorCode;
import com.cinema.user.dto.*;
import com.cinema.user.entity.User;
import com.cinema.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(readOnly = true)
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public UserResponse getUserById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));
        return toUserResponse(user);
    }

    public UserResponse getUserByUsername(String username) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));
        return toUserResponse(user);
    }

    @Transactional
    public UserResponse updateProfile(Long userId, UpdateProfileRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

        if (request.getFullName() != null) {
            user.setFullName(request.getFullName());
        }
        if (request.getPhoneNumber() != null) {
            user.setPhoneNumber(request.getPhoneNumber());
        }
        if (request.getAvatarUrl() != null) {
            user.setAvatarUrl(request.getAvatarUrl());
        }

        user = userRepository.save(user);
        log.info("User {} updated profile", userId);
        return toUserResponse(user);
    }

    @Transactional
    public void changePassword(Long userId, ChangePasswordRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

        if (!passwordEncoder.matches(request.getCurrentPassword(), user.getPasswordHash())) {
            throw new BusinessException(ErrorCode.AUTH_INVALID_CREDENTIALS, "Current password is incorrect");
        }

        user.setPasswordHash(passwordEncoder.encode(request.getNewPassword()));
        user.incrementTokenVersion(); // Invalidate all existing tokens
        userRepository.save(user);
        log.info("User {} changed password", userId);
    }

    // Admin methods
    public Page<UserResponse> getAllUsers(Pageable pageable) {
        return userRepository.findAll(pageable).map(this::toUserResponse);
    }

    @Transactional
    public UserResponse adminUpdateUser(Long userId, AdminUpdateUserRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));

        if (request.getFullName() != null) {
            user.setFullName(request.getFullName());
        }
        if (request.getPhoneNumber() != null) {
            user.setPhoneNumber(request.getPhoneNumber());
        }
        if (request.getRole() != null) {
            user.setRole(User.UserRole.valueOf(request.getRole()));
        }
        if (request.getStatus() != null) {
            user.setStatus(User.UserStatus.valueOf(request.getStatus()));
        }
        if (request.getPoints() != null) {
            user.setPoints(request.getPoints());
        }

        user = userRepository.save(user);
        log.info("Admin updated user {}", userId);
        return toUserResponse(user);
    }

    @Transactional
    public void addPoints(Long userId, int points) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));
        user.addPoints(points);
        userRepository.save(user);
        log.info("Added {} points to user {}", points, userId);
    }

    @Transactional
    public void deductPoints(Long userId, int points) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ErrorCode.USER_NOT_FOUND));
        if (!user.deductPoints(points)) {
            throw new BusinessException(ErrorCode.USER_INSUFFICIENT_POINTS);
        }
        userRepository.save(user);
        log.info("Deducted {} points from user {}", points, userId);
    }

    private UserResponse toUserResponse(User user) {
        return UserResponse.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .phoneNumber(user.getPhoneNumber())
                .avatarUrl(user.getAvatarUrl())
                .points(user.getPoints())
                .role(user.getRole().name())
                .status(user.getStatus().name())
                .lastLoginAt(user.getLastLoginAt())
                .createdAt(user.getCreatedAt())
                .build();
    }
}

File: frontend/src/api/user.service.ts
import axiosInstance from './axios';
import type {
  UserResponse,
  UpdateProfileRequest,
  ChangePasswordRequest,
  AdminUpdateUserRequest,
  ApiResponse,
  PaginatedResponse,
} from '@/types';

/**
 * User Service
 * Handles all user-related API calls
 */
export const userService = {
  /**
   * GET /api/users/profile
   * Get current user profile
   */
  getProfile: async (): Promise<UserResponse> => {
    const response = await axiosInstance.get<ApiResponse<UserResponse>>(
      '/users/profile'
    );
    return response.data.data;
  },

  /**
   * PUT /api/users/profile
   * Update user profile
   */
  updateProfile: async (data: UpdateProfileRequest): Promise<UserResponse> => {
    const response = await axiosInstance.put<ApiResponse<UserResponse>>(
      '/users/profile',
      data
    );
    return response.data.data;
  },

  /**
   * PUT /api/users/change-password
   * Change password
   */
  changePassword: async (data: ChangePasswordRequest): Promise<void> => {
    await axiosInstance.put('/users/change-password', data);
  },

  /**
   * GET /api/users/{id}
   * Get user by ID (admin)
   */
  getUserById: async (id: number): Promise<UserResponse> => {
    const response = await axiosInstance.get<ApiResponse<UserResponse>>(
      `/users/${id}`
    );
    return response.data.data;
  },

  /**
   * GET /api/users/username/{username}
   * Get user by username (admin)
   */
  getUserByUsername: async (username: string): Promise<UserResponse> => {
    const response = await axiosInstance.get<ApiResponse<UserResponse>>(
      `/users/username/${username}`
    );
    return response.data.data;
  },

  /**
   * GET /api/users
   * Get all users with pagination (admin)
   */
  getAllUsers: async (params: {
    page?: number;
    size?: number;
  }): Promise<PaginatedResponse<UserResponse>> => {
    const response = await axiosInstance.get<
      ApiResponse<PaginatedResponse<UserResponse>>
    >('/users', { params });
    return response.data.data;
  },

  /**
   * PUT /api/users/{id}
   * Admin update user
   */
  adminUpdateUser: async (
    id: number,
    data: AdminUpdateUserRequest
  ): Promise<UserResponse> => {
    const response = await axiosInstance.put<ApiResponse<UserResponse>>(
      `/users/${id}`,
      data
    );
    return response.data.data;
  },

  /**
   * POST /api/users/{id}/add-points
   * Add points to user (admin)
   */
  addPoints: async (id: number, points: number): Promise<void> => {
    await axiosInstance.post(`/users/${id}/add-points`, { points });
  },

  /**
   * POST /api/users/{id}/deduct-points
   * Deduct points from user (admin)
   */
  deductPoints: async (id: number, points: number): Promise<void> => {
    await axiosInstance.post(`/users/${id}/deduct-points`, { points });
  },

  /**
   * POST /api/users/{id}/lock
   * Lock user account (admin)
   */
  lockUser: async (id: number): Promise<void> => {
    await axiosInstance.post(`/users/${id}/lock`);
  },

  /**
   * POST /api/users/{id}/unlock
   * Unlock user account (admin)
   */
  unlockUser: async (id: number): Promise<void> => {
    await axiosInstance.post(`/users/${id}/unlock`);
  },

  /**
   * PUT /api/admin/users/{id}/status?status=ACTIVE
   * Update user status (ACTIVE, INACTIVE, BANNED) (admin)
   */
  updateUserStatus: async (
    id: number,
    status: string
  ): Promise<UserResponse> => {
    const response = await axiosInstance.put<ApiResponse<UserResponse>>(
      `/admin/users/${id}/status`,
      null,
      { params: { status } }
    );
    return response.data.data;
  },

  /**
   * PUT /api/admin/users/{id}/role?role=ADMIN
   * Update user role (CUSTOMER, STAFF, ADMIN) (admin)
   */
  updateUserRole: async (id: number, role: string): Promise<UserResponse> => {
    const response = await axiosInstance.put<ApiResponse<UserResponse>>(
      `/admin/users/${id}/role`,
      null,
      { params: { role } }
    );
    return response.data.data;
  },
};
